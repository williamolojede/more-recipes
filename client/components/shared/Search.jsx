import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import instance from '../../config/axios';

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchResults: []
    };
  }


  componentDidMount() {
    this.searchInput.focus();
  }

  search = () => {
    const searchString = this.searchInput.value;
    instance.get(`/recipes?search=${searchString}`)
      .then((res) => {
        this.setState({
          searchResults: res.data.recipes
        });
      });
  }

  render() {
    const { searchResults } = this.state;
    const { toggleSearch } = this.props;
    return (
      <div className="search">

        <div className="search__container">

          <div className="search__top">
            <p>Search</p>
            <button className="search__cancel-button" onClick={toggleSearch}>
              <i className="material-icons">clear</i>
            </button>
          </div>

          <textarea
            ref={(node) => { this.searchInput = node; }}
            onChange={e => this.search(e)}
          />

          <ul className="search__results">
            {
              searchResults.map(result => (
                <li className="search__resuls-item" key={result.id}>
                  <Link to={`/recipe/${result.id}`}>{result.name}</Link>
                </li>
              ))
            }
          </ul>

        </div>
      </div>
    );
  }
}

Search.propTypes = {
  toggleSearch: PropTypes.func.isRequired
};

export default Search;
