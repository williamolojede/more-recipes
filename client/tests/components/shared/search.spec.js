import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import Search from '../../../components/shared/Search';
import recipesMock from '../../__mocks__/recipes.mock';

jest.useFakeTimers();

function setup() {
  const props = {
    toggleSearch: jest.fn()
  };
  const wrapper = mount(<Router><Search {...props} /></Router>);

  return {
    wrapper
  };
}

describe('Search', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a list of recipe names that match search term', () => {
    const { wrapper } = setup();
    const search = wrapper.find('Search');
    search.instance().setState({ searchResults: recipesMock.recipes });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a list of recipe names that match search term', () => {
    const { wrapper } = setup();
    const search = wrapper.find('Search');
    const searchSpy = jest.spyOn(search.instance(), 'search');
    const textarea = wrapper.find('textarea');
    expect(textarea.length).toBe(1);
    textarea.simulate('change', eventObject);
    jest.runAllTimers();
    expect(searchSpy).toHaveBeenCalledTimes(1);
  });
});
