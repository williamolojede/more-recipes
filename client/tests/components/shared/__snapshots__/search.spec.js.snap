// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render a list of recipe names that match search term 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "blank",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Search
      toggleSearch={[Function]}
    >
      <div
        className="search"
      >
        <div
          className="search__container"
        >
          <div
            className="search__top"
          >
            <p>
              Search
            </p>
            <button
              className="search__cancel-button"
              onClick={[Function]}
            >
              <i
                className="material-icons"
              >
                clear
              </i>
            </button>
          </div>
          <textarea
            onChange={[Function]}
          />
          <ul
            className="search__results"
          />
        </div>
      </div>
    </Search>
  </Router>
</BrowserRouter>
`;

exports[`Search should render with right amount of elements 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "blank",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Search
      toggleSearch={[Function]}
    >
      <div
        className="search"
      >
        <div
          className="search__container"
        >
          <div
            className="search__top"
          >
            <p>
              Search
            </p>
            <button
              className="search__cancel-button"
              onClick={[Function]}
            >
              <i
                className="material-icons"
              >
                clear
              </i>
            </button>
          </div>
          <textarea
            onChange={[Function]}
          />
          <ul
            className="search__results"
          />
        </div>
      </div>
    </Search>
  </Router>
</BrowserRouter>
`;
